### PR TITLE
Fix partition info in segment ZK metadata for changes in stream partition count

### PR DESCRIPTION
**Pull request by sajjad-moradi**
## Description
This PR handles the bug described in issue #11475 .

## Testing Done
Verified the correct behavior by changing `LLCRealtimeClusterIntegrationTest`:
- created a Kafka topic with 2 partitions
- created a table with partitioning enabled (partition count was set as 6 to not match the stream partition count)
- published 4 events with id's [0, 1, 2, 3] to the topic
- verified that partitioning info in segment ZK metadata is correctly set based on 2 partitions, not 6
- increased the Kafka topic partition count to 4
- published 4 more events with id's [4, 5, 6, 7]
- at this point, there's only two consuming segments. Some of the new events went to the existing consuming segments, for some there's no consuming segment available.
- kicked off `RealtimeSegmentValidationManager` periodic job to create new consuming segments
- verified that for the new consuming segments, `numPartition` field in segment ZK metadata is correctly set as 4, while for the existing consuming, the value is 2.
- queried the table with different id's in where clause to make sure that partition pruning / routing works as expected
- issued a force-commit
- verified that partition info for committed segments as well as new consuming segments are correctly set 

Note: I think it's best if later we put all the above steps in a separate integration test, so we make sure the expected behavior doesn't change with future refactoring.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 1]
```diff
@@ -719,8 +720,13 @@ private SegmentPartitionMetadata getPartitionMetadataFromTableConfig(TableConfig
     if (columnPartitionMap.size() == 1) {
       Map.Entry<String, ColumnPartitionConfig> entry = columnPartitionMap.entrySet().iterator().next();
       ColumnPartitionConfig columnPartitionConfig = entry.getValue();
+      if (numPartitionGroups != columnPartitionConfig.getNumPartitions()) {
+        LOGGER.warn("Number of partition groups fetched from the stream is different than "
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:724]: Should we also include both numbers in the log for debugging purpose?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM

**Review by INNOCENT-BOY — APPROVED** [review]
LGTM

**Review by INNOCENT-BOY — APPROVED** [review]
LGTM

**Comment by jadami10** [comment]
@sajjad-moradi, is this supposed to fix consuming segments? we'd seen this before, but when going from CONSUMING -> COMPLETED, the right partitions were then stored in the ZKMetadata

**Comment by sajjad-moradi** [comment]
> @sajjad-moradi, is this supposed to fix consuming segments? we'd seen this before, but when going from CONSUMING -> COMPLETED, the right partitions were then stored in the ZKMetadata

Yes, this is for consuming segments. It usually takes hours (depending on stream configs) for consuming segment to start committing. Without the fix, consuming segments for new partitions have wrong ZK metadata => which results in incorrect partition routing => which results in incorrect query responses

**Comment by jadami10** [comment]
This only works when the new segments are created, right? Consuming segments during repartition will still be incorrect until synced?

**Comment by sajjad-moradi** [comment]
This is for fixing the metadata of the new consuming segments for the new partitions. The consuming segments for existing partitions keep their existing partitioning metadata. Depending on how you repartition, that might be correct or incorrect. For example, in Linkedin, any partition change is a factor of two. With this repartitioning approach, the existing segment metadata will result in correct partition routing. 

Let say there are two partitions and 4 ids in total for the partitioning column. So before repartitioning, here are segment ZK metadata for the two partitions:
p0 -> numPartitions: 2, partitionIds = [0] (corresponding to id's 0, 2)
p1 -> numPartitions: 2, partitionIds = [1] (corresponding to id's 1, 3)

After repartitioning, two new partitions are added, but the existing consuming segment keep their existing metadata:
p0 -> numPartitions: 2, partitionIds = [0] (corresponding to id's 0, 2)
p1 -> numPartitions: 2, partitionIds = [1] (corresponding to id's 1, 3)
p2 -> numPartitions: 4, partitionIds = [2] (corresponding to id 2)
p3 -> numPartitions: 4, partitionIds = [3] (corresponding to id 3)

After the next segment commits, consuming segment zk metadata are like these:
p0 -> numPartitions: 4, partitionIds = [0] (corresponding to id 0)
p1 -> numPartitions: 4, partitionIds = [1] (corresponding to id 1)
p2 -> numPartitions: 4, partitionIds = [2] (corresponding to id 2)
p3 -> numPartitions: 4, partitionIds = [3] (corresponding to id 3)

This shows that with "factor or two" repartitioning approach, the partition routing is correct before, during, and after repartitioning. 

If you're using different repartitioning, you need to do a force-commit right after repartition happens. This way you can significantly reduce the duration when partition aware routing is problematic.

**Pull request merged**